File: 2_trees/1_Balanced_Binary_Tree.cpp
// Balanced Binary Tree

/*
Given a binary tree, determine if it is height-balanced.

For this problem, a height-balanced binary tree is defined as:

a binary tree in which the depth of the two subtrees of every node never differ by more than 1.

Example 1:

Given the following tree [3,9,20,null,null,15,7]:

    3
   / \
  9  20
    /  \
   15   7
Return true.

Example 2:

Given the following tree [1,2,2,3,3,null,null,4,4]:

       1
      / \
     2   2
    / \
   3   3
  / \
 4   4
Return false.
*/

// my solution - Not optimal
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    int height(TreeNode* node) {
        if (node == NULL)
            return 0;
        
        return 1 + max(height(node->left), height(node->right));
    }
public:
    bool isBalanced(TreeNode* root) {
        int lh, rh;
        
        if (root == NULL)
            return true;
        
        lh = height(root->left);
        rh = height(root->right);
        
        if ((abs(lh - rh) <= 1) && isBalanced(root->left) && isBalanced(root->right))
            return true;
        
        return false;
    }
};

File: 2_trees/1_Balanced_Binary_Tree.py
"""
Balanced Binary Tree

Solution
Given a binary tree, determine if it is height-balanced.

For this problem, a height-balanced binary tree is defined as:

a binary tree in which the left and right subtrees of every node differ in height by no more than 1.

Example 1:

Input: root = [3,9,20,null,null,15,7]
Output: true
Example 2:


Input: root = [1,2,2,3,3,null,null,4,4]
Output: false
Example 3:

Input: root = []
Output: true

Constraints:

The number of nodes in the tree is in the range [0, 5000].
-104 <= Node.val <= 104
"""
# Optimal O(n) time, O(1) space, O(log(n)) stack

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    
    def isBalanced(self, root: TreeNode) -> bool:
        if root is None:
            return True
        
        self.result = True
        self.checkBalanced(root, 0)
        
        return self.result
    
    def checkBalanced(self, root: TreeNode, depth: int) -> bool:
        if root is None or self.result == False:
            return depth
        
        l = self.checkBalanced(root.left, depth + 1)
        r = self.checkBalanced(root.right, depth + 1)
        
        if abs(l-r) > 1:
            self.result = False
        
        return max(l, r)